### PR TITLE
Capture FAULT exception

### PR DESCRIPTION
**Pull request by shargon**
Close https://github.com/neo-project/neo-vm/issues/331

### CODE REVIEW

**Comment by shargon** [comment]
But if we set FAULT first, it won't be used, did you prefer a new variable?

**Comment by erikzhang** [comment]
No. I prefer not to save these implementation-specific data.

**Comment by shargon** [comment]
Please, check it again @erikzhang

**Comment by erikzhang** [comment]
Can you clarify where this will be used?

**Comment by shargon** [comment]
We can log it if we are under debug, or show a hint of where the FAULT is

**Comment by erikzhang** [comment]
> if (!ExecuteInstruction(instruction)) OnFault(new Exception(nameof(ExecuteInstruction)));

But this information is useless. If we can return the actual exception, then it will be better.

**Comment by Tommo-L** [comment]
> Can you clarify where this will be used?

When user invoke the `invokescript` of rpc, or send a transaction in neo-cli or sdk, the script may execute failure, it'll be better to show the error message.   Here is an example, user even don't know what happened only `FAULT` status. https://github.com/neo-project/neo-node/issues/600

**Comment by Tommo-L** [comment]
> But this information is useless. 

The exception throwed in neo-vm maybe useless, but the exception throwed in neo-core is useful.

**Comment by erikzhang** [comment]
> When user invoke the invokescript of rpc, or send a transaction in neo-cli or sdk, the script may execute failure, it'll be better to show the error message. Here is an example, user even don't know what happened only FAULT status.

Isn't it better to show users the contents of `InvocationStack` and `EvaluationStack`?

**Comment by erikzhang** [comment]
Do you think that `PreExecuteInstruction()`, `ExecuteInstruction()`, and `PostExecuteInstruction()` should not return `bool`, but should use exceptions to put the VM into `FAULT` state?

**Comment by shargon** [comment]
> Do you think that PreExecuteInstruction(), ExecuteInstruction(), and PostExecuteInstruction() should not return bool, but should use exceptions to put the VM into FAULT state?

Throw exception it's slower, so I prefer a dirty solution

```csharp
bool PreExecuteInstruction(out string error);
```

**Comment by erikzhang** [comment]
Although it is slow, but it will happen only once for each `ExecutionEngine`. So I think it's okay with exceptions. The performance loss is negligible.

**Comment by Tommo-L** [comment]
> Although it is slow, but it will happen only once for each ExecutionEngine.

That's right. For me it's okay, we don't need so many `return false`, we can just throw exception in `pop` and `peek` methods.

https://github.com/neo-project/neo-vm/blob/01d6776862a195019f51a8096e7b68d3a2052090/src/neo-vm/ExecutionEngine.cs#L212-L216

**Comment by shargon** [comment]
Then I will change this PR to throw the right exception

**Comment by shargon** [comment]
> That's right. For me it's okay, we don't need so many return false, we can just throw exception in pop and peek methods.

I prefer to throw exception instead of `return false`, change TryPop, it's not needed, what do you think @erikzhang ?

**Comment by erikzhang** [comment]
We can remove `TryPop()`.

**Comment by shargon** [comment]
> We can remove TryPop().

Then we back to pop items if it fault.

**Comment by shargon** [comment]
Please take a look if you mean like this https://github.com/shargon/neo-vm/pull/3 in order to fix the ut and change `ExecuteInstruction` to ```private void ExecuteInstruction(Instruction instruction)```

**Comment by erikzhang** [comment]
@shargon You can't change `break` to `return`. Because there are extra steps to be run at the end of `switch`.

```c#
context.MoveNext();
```

**Comment by erikzhang** [comment]
Sorry, I just saw that is for JUMPs.

File: src/neo-vm/ExecutionEngine.cs [hunk 1]
```diff
@@ -1338,11 +1404,19 @@ public ExecutionContext LoadScript(Script script, int initialPosition = 0)
             return context;
         }
 
+        protected virtual void OnFault(Exception e)
```
Comment (Tommo-L) [inline src/neo-vm/ExecutionEngine.cs:1407]: @shargon  @erikzhang  Could we capture exception in neo-core now?

**Pull request merged**